examples/admin-partitions/ap-example

```diff
@@ -0,0 +1,74 @@
+#!/bin/bash
+
+waitfor() {
+  echo -n "waiting for ${1}/${2}/${3} to be registered in Consul"
+  local count=0
+  while [[ $count -le 30 ]]; do
+    echo -n "."
+    echo $(curl -sS -X GET \
+      -H "Authorization: Bearer ${CONSUL_HTTP_TOKEN}"\
+      "${CONSUL_HTTP_ADDR}/v1/catalog/services?partition=${1}&ns=${2}") \
+      | grep -q "${3}" && return
+    sleep 10
+    ((count++))
+  done
+  echo ""
+  echo "timeout waiting for ${1}/${2}/${3}"
+}
+
+# get Terraform outputs and initialize required variables.
+echo "loading Terraform outputs"
+tfOutputs=$(cd terraform ; terraform output -json)
+CONSUL_HTTP_ADDR=$(echo "$tfOutputs" | jq -rc '.hcp_public_endpoint.value')
+CONSUL_HTTP_TOKEN=$(echo "$tfOutputs" | jq -rc '.token.value')
+
+CLIENT_REGION=$(echo "$tfOutputs" | jq -rc '.client.value.region')
+CLIENT_CLUSTER_ARN=$(echo "$tfOutputs" | jq -rc '.client.value.ecs_cluster_arn')
+CLIENT_NAME=$(echo "$tfOutputs" | jq -rc '.client.value.name')
+CLIENT_AP=$(echo "$tfOutputs" | jq -rc '.client.value.partition')
+CLIENT_NS=$(echo "$tfOutputs" | jq -rc '.client.value.namespace')
+
+SERVER_REGION=$(echo "$tfOutputs" | jq -rc '.server.value.region')
+SERVER_CLUSTER_ARN=$(echo "$tfOutputs" | jq -rc '.server.value.ecs_cluster_arn')
+SERVER_NAME=$(echo "$tfOutputs" | jq -rc '.server.value.name')
+SERVER_AP=$(echo "$tfOutputs" | jq -rc '.server.value.partition')
+SERVER_NS=$(echo "$tfOutputs" | jq -rc '.server.value.namespace')
+
+# wait for services to be registered in Consul
+waitfor ${CLIENT_AP} ${CLIENT_NS} ${CLIENT_NAME}
+echo ""
+
+waitfor ${SERVER_AP} ${SERVER_NS} ${SERVER_NAME}
+echo ""
+
+# create exported-services config entry to expose the example_server in partition 2 to the services in partition 1
+echo "creating exported-services config entry"
```

Comment:
Both are good suggestions. Long term I like the idea of using the Terraform consul provider to do this since I think that is the best practicable example we can provide.

The example, as it currently is, is a bit contrived and probably not how things would be done in practice. I did this to test that the ACL controller was doing specific things like creating the partition and namespace(s) for the tasks if they don't already exist. But those types of checks belong in the acceptance tests (which they do now), not an example.

> Actually, it looks like intentions are managed via config entries as well (TIL)

Yes, I learned that recently too and I updated the HCP acceptance tests to use config-entry-style intentions.

Reply:
The Consul TF provider does not support admin partitions. I updated the `hcp_consul.tf` file with the proposed updates and have commented them out with a TODO for now. I have added a task to the backlog to update this once the TF provider supports admin partitions.

I am going to leave the script as is (using raw HTTP requests) for now since I think the TF provider is the way to go and I don't want to update this thrice.